### PR TITLE
Transmit additional information when remote handler execution failed

### LINKED ISSUE #7017: Improve log messages when communication failed
**Description**

The log messages:
* "No remote message handler registered for this message" https://console.cloud.google.com/errors/COeS4vn5-Je6jwE?service=zeebe-gateway&version=medic-cw-18-953a27a262-benchmark&time=P7D&project=zeebe-io
* "Remote handler failed to handle message" https://console.cloud.google.com/errors/CKT115CooYK3_wE?service=zeebe-gateway&version=medic-cw-18-953a27a262-benchmark&time=P7D&project=zeebe-io

Are not very helpful.

**Expected Behavior**
The payload is used to transmit information that can be used to enhance the log message.

E.g.: 
* "no remote message handler" - transmit subject of original message
* "failed to handle messge" - transmit first lines of stacktrace, or - if that is too much information - at least the exception message

### PR DESCRIPTION
**Pull request by romansmirnov**
## Description

* `RemoteHandlerFailure`: Transmit the entire exception message when the execution of the handler fails, so that the sender logs contain the exception message as well.
* `NoRemoteHandler`: Transmit subject when a remote handler is not found

## Related issues

closes #7017 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Comment by romansmirnov** [comment]
@npepinpe,

This is my first PR but it is not ready for review yet - I need to write another test case to cover the second handler registration.

In terms of test cases, I decided to write the test cases in `NettyMessagingServiceTest` which allowed me to simulate the situation when the execution of a handler fails. Are there any other options to write a test case?

**Comment by npepinpe** [comment]
Cool, just let me know when it's ready for review.

As for your question, I'm not sure what you mean. Do you mean in terms of unit/integration testing? I think for your particular issue, testing in the `NettyMessagingServiceTest` is fine, as you're not targeting a specific usage of the messaging service (although feel free to challenge this if you think otherwise :slightly_smiling_face:).

You could do even tighter unit testing of course by mocking the other service, but I don't think we would gain much here since it's already relatively fast, and it's fairly easy to diagnose failures.

File: atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java [hunk 1]
```diff
@@ -429,4 +432,214 @@ public void shouldNotBindToAdvertisedAddress() {
     assertThat(messagingService.bindingAddresses()).contains(bindingAddress);
     assertThat(messagingService.address()).isEqualTo(nonBindableAddress);
   }
+
+  @Test
+  public void testRemoteHandlerFailure() {
+    // given
+    final String exceptionMessage = "foo bar";
+    final RemoteHandlerFailure execptedException =
```
Comment (npepinpe) [inline atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java:440]: :wrench: Feel free to use `var` keyword when the type is obvious from the assignment. For example, here it's pretty clear, but lower when declaring a `BiFunction` it's not, so it makes sense to keep it. However, we don't have a strong guideline on this and it's more of a suggestion.

File: atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java [hunk 2]
```diff
@@ -429,4 +432,214 @@ public void shouldNotBindToAdvertisedAddress() {
...
+
+    // then
+    try {
+      response.join();
+      fail("MessagingException.RemoteHandlerFailure expected");
+    } catch (Exception e) {
+      final Throwable cause = e.getCause();
+      assertTrue(cause instanceof MessagingException.RemoteHandlerFailure);
+      assertTrue(execptedException.getMessage().equals(cause.getMessage()));
+    }
```
Comment (npepinpe) [inline atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java:463]: :x: You'll notice that the forked Atomix code still uses some junit and Hamcrest assertions. For new code, or when you update that code, we always use AssertJ. So here, for example, you could use AssertJ's `Future`/`CompletableFuture` specific assertions, e.g.

```java
assertThat(response)
    .failsWithin(Duration.ofSeconds(30))
    .isInstanceOf(ExecutionException.class)
    .havingRootCause()
    .isInstanceOf(MessagingException.RemoteHandlerFailure)
    .havingMessage(cause.getMessage);
```

(I didn't test this code though, so maybe the method names aren't exactly right :sweat_smile:, but it should give you an idea)

Comment (npepinpe) [reply atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java:463]: The same applies to the other tests below.

**Review by npepinpe — CHANGES_REQUESTED** [review]
:rocket: 

Thanks Roman! I had some notes, but nothing major, just one small point, and I know it's one that's impossible to know without someone telling you :smile: You'll notice quickly we talk about Atomix and Zeebe code, and while we've hard forked Atomix and it's now effectively Zeebe (:sweat_smile:), large chunks of it weren't updated, so they sometimes use different conventions than in our code. The rule of thumb is, if you have to touch the old code, update it to how we do it in Zeebe (provided it doesn't blow up the scope of your issue, e.g. migrating Atomix to use actors :smile:), and for new code we do it as we do for the rest of our "Zeebe" code.

So in this case, for example, Atomix used junit and Hamcrest for assertions. We always use AssertJ for our own code, so when writing new tests, or updating Atomix tests, we'll use that as well.

I'll go more in depth in the differences between Atomix, Zeebe, and our long term plans there, in our 1:1s. In the meantime, you can always ask the reviewer or anyone when you kick off an issue, or just ask me directly at any point if you're unsure or something seems inconsistent between different parts of the code base, and what would be the consistent way moving forward.

File: atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java (hunk 1 above)
Comment (romansmirnov) [reply atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java]: Thanks! I applied the requested changes.

File: atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java (hunk 2 above)
Comment (romansmirnov) [reply atomix/cluster/src/test/java/io/atomix/cluster/messaging/impl/NettyMessagingServiceTest.java]: Thanks for the hint. I applied the requested changes.

**Review by npepinpe — APPROVED** [review]
:tada: 

One small thing, I know it's a little out of scope, but when I was running the tests you wrote I wanted to correlate what I saw between the services, and noticed the logging on the handler side is kind of poor on the `RemoteHandlerFailure` case. I know you didn't write it, and it's not part of the original issue, but I would propose making this tiny change here as well if we're touching that code, i.e. following the boy scout rule.

Right now, it logs the following:

```
13:51:02.204 [] WARN  io.atomix.cluster.messaging.impl.NettyMessagingService - An error occurred in a message handler:
java.lang.RuntimeException: foo bar
	at io.atomix.cluster.messaging.impl.NettyMessagingServiceTest.lambda$testCompletableRemoteHandlerFailure$17(NettyMessagingServiceTest.java:524) ~[test-classes/:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$registerHandler$11(NettyMessagingService.java:281) ~[classes/:?]
	at io.atomix.cluster.messaging.impl.AbstractServerConnection.dispatch(AbstractServerConnection.java:40) [classes/:?]
	at io.atomix.cluster.messaging.impl.AbstractServerConnection.dispatch(AbstractServerConnection.java:26) [classes/:?]
```

I would propose changing it to:

```
13:54:32.702 [] WARN  io.atomix.cluster.messaging.impl.NettyMessagingService - Unexpected error while handling message 5aadd4e3-df4b-4a83-9418-19a032dea301 from bigfoot:1025
java.lang.RuntimeException: foo bar
	at io.atomix.cluster.messaging.impl.NettyMessagingServiceTest.lambda$testCompletableRemoteHandlerFailure$17(NettyMessagingServiceTest.java:524) ~[test-classes/:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService.lambda$registerHandler$11(NettyMessagingService.java:281) ~[classes/:?]
	at io.atomix.cluster.messaging.impl.AbstractServerConnection.dispatch(AbstractServerConnection.java:40) [classes/:?]
	at io.atomix.cluster.messaging.impl.AbstractServerConnection.dispatch(AbstractServerConnection.java:26) [classes/:?]
	at io.atomix.cluster.messaging.impl.NettyMessagingService$MessageDispatcher.channelRead0(NettyMessagingService.java:1001) [classes/:?]
```

Effectively adding the subject and the client which triggered this error. This should help us correlate errors on both sides. This means replacing 

```java
                        log.warn("An error occurred in a message handler:", error);
```

with

```java
                        log.warn(
                            "Unexpected error while handling message {} from {}",
                            message.subject(),
                            message.sender(),
                            error);
```

Let me know what you think, otherwise I can open a follow up PR myself. I'm approving anyway as I don't consider this a blocker :slightly_smiling_face:

**Comment by romansmirnov** [comment]
@npepinpe, your suggestion makes sense to me. I will adjust the exception message.

**Comment by romansmirnov** [comment]
@npepinpe, I applied your last suggestion and the build runs through so I will merge the changes.

**Comment by romansmirnov** [comment]
bors merge

**Pull request merged**